rts/lra/lra-proxy/api/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRAParticipant.java

```diff
@@ -0,0 +1,209 @@
+import javax.ws.rs.core.UriBuilder;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+
+import static io.narayana.lra.LRAConstants.COMPENSATE;
+import static io.narayana.lra.LRAConstants.COMPLETE;
+import static io.narayana.lra.LRAConstants.FORGET;
+import static io.narayana.lra.LRAConstants.STATUS;
+
+/**
+ * Keeps references to individual non-JAX-RS paraticipant methods in
+ * single LRA participant class.
+ */
+public class LRAParticipant {
+
+    private Class<?> id;
+    private Method compensateMethod;
+    private Method completeMethod;
+    private Method statusMethod;
+    private Method forgetMethod;
+
+    LRAParticipant(Class<?> id) {
+        this.id = id;
+    }
+
+    Class<?> getId() {
+        return id;
+    }
+
+    Response compensate(URI lraId, URI parentId) {
+        return invokeParticipantMethod(compensateMethod, lraId, parentId, COMPENSATE);
+    }
+
+    Response complete(URI lraId, URI parentId) {
+        return invokeParticipantMethod(completeMethod, lraId, parentId, COMPLETE);
+    }
+
+    Response status(URI lraId, URI parentId) {
+        return invokeParticipantMethod(statusMethod, lraId, parentId, STATUS);
+    }
+
+    Response forget(URI lraId, URI parentId) {
+        return invokeParticipantMethod(forgetMethod, lraId, parentId, FORGET);
+    }
+
+    @SuppressWarnings("unchecked")
+    private Response invokeParticipantMethod(Method method, URI lraId,
+                                             URI parentId, String type) {
+        Object participant = CDI.current().select(id).get();
+        Object result = null;
+
+        switch (method.getParameterCount()) {
+            case 0:
+                result = invokeMethod(type, method, participant);
+                break;
+            case 1:
+                result = invokeMethod(type, method, participant, lraId);
+                break;
+            case 2:
+                result = invokeMethod(type, method, participant, lraId, parentId);
+                break;
+            default:
+                throw new InvalidLRAParticipantDefinitionException(
+                    method.toGenericString() + ": invalid number of arguments: " + method.getParameterCount());
+        }
+
+        Response.ResponseBuilder builder = Response.status(Response.Status.OK);
+
+        if (result instanceof CompletionStage) {
+            // TODO optimize this with coordinator
+            // wait for result the same way as JAX-RS filters do
+            try {
+                result = ((CompletionStage) result).toCompletableFuture().get();
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        if (result == null) {
+            builder.status(Response.Status.NOT_FOUND);
```

Comment:
result will be null if the return type of the method is void so returning NOT_FOUND is incorrect.